### PR TITLE
Feature/performance improvements

### PR DESCRIPTION
**Pull request by michielvandergeest**

### CODE REVIEW

File: src/engines/L3/element.js [hunk 1]
```diff
@@ -102,9 +101,9 @@ const propsTransformer = {
     this.props['zIndex'] = v
   },
   set color(v) {
-    if (typeof v === 'string' && v.indexOf('{') === -1) {
+    if (typeof v === 'string' && v.startsWith('{') === false) {
```
Comment (wouterlucas) [inline src/engines/L3/element.js:104]: nice one ❤️

File: src/engines/L3/element.js [hunk 2]
```diff
@@ -266,65 +258,63 @@ const Element = {
...
+    if (this.props.props['color'] === undefined && '__textnode' in props === false) {
       this.props.props['color'] = 0
     }
 
     this.node = props.__textnode
       ? renderer.createTextNode({ ...textDefaults, ...this.props.props })
       : renderer.createNode(this.props.props)
 
-    if (props['@loaded']) {
+    if ('@loaded' in props === true && typeof props['@loaded'] === 'function') {
```
Comment (wouterlucas) [inline src/engines/L3/element.js:269]: 👀 

is `'@loaded' in props === true` faster then an undefined check?

File: src/lib/codegenerator/generator.js [hunk 3]
```diff
@@ -287,7 +306,12 @@ const generateForLoopCode = function (templateObject, parent) {
     const created${forStartCounter} = []
     const forloop${forStartCounter} = (collection = [], elms, created) => {
       const rawCollection = getRaw(collection)
-      const keys = new Set(rawCollection.map((${item}) => '' +  ${interpolate(key, '')}))
+      const keys = new Set()
+      let l = rawCollection.length
+      while(l--) {
```
Comment (wouterlucas) [inline src/lib/codegenerator/generator.js:311]: ❤️  while loops for the win

File: src/lib/colors/colors.js [hunk 4]
```diff
@@ -39,17 +39,16 @@ export default {
...
     if (color.startsWith('0x') && color.length === 10) {
-      colorsMap[color] = color
       return color
     }
 
+    // it is a previously mapped color, return it
+    if (colorsMap[color] !== undefined) {
+      return colorsMap[color]
+    }
+
```
Comment (wouterlucas) [inline src/lib/colors/colors.js:51]: this means you're no longer caching `0x` hex colors, is `(color.startsWith('0x') && color.length === 10)` faster then `(colorsMap[color] !== undefined)`?

**Review by wouterlucas — APPROVED** [review]
LGTM just some Qs for learnings 👍

File: src/lib/colors/colors.js (hunk 4 above)
Comment (michielvandergeest) [reply src/lib/colors/colors.js:51]: Correct, the difference between the 2 check is close to negligible. In the case of `0x` colors it doesn't really make sense to cache the result as the same value, without manipulation, will be returned anyway. So with this change we don't store data unnecessarily. For most normal app scenarios this won't be a problem, but when working with random colors in test scenarios the cache can fill up.

File: src/engines/L3/element.js (hunk 2 above)
Comment (michielvandergeest) [reply src/engines/L3/element.js:269]: No it's not :smile: .. changing to `undefined` check

File: src/lib/colors/colors.js (hunk 4 above)
Comment (wouterlucas) [reply src/lib/colors/colors.js:51]: Fair enough - thanks for explaining!

**Pull request merged**